var/spack/repos/builtin/packages/sz/package.py

```diff
@@ -32,17 +30,73 @@ class Sz(AutotoolsPackage):
     version('1.4.10.0', sha256='cf23cf1ffd7c69c3d3128ae9c356b6acdc03a38f92c02db5d9bfc04f3fabc506')
     version('1.4.9.2',  sha256='9dc785274d068d04c2836955fc93518a9797bfd409b46fea5733294b7c7c18f8')
 
-    variant('fortran', default=False,
-            description='Enable fortran compilation')
+
+    variant('python', default=False, description="builds the python wrapper")
+    variant('netcdf', default=False, description="build the netcdf reader")
+    variant('hdf5', default=False, description="build the hdf5 filter")
+    variant('pastri', default=False, description="build the pastri mode")
+    variant('time_compression', default=False, description="build the time based compression mode")
+    variant('random_access', default=False, description="build the random access compression mode")
+    variant('fortran', default=False, description='Enable fortran compilation')
+    variant('shared', default=True, description="build shared versions of the libraries")
 
     # Part of latest sources don't support -O3 optimization
     # with Fujitsu compiler.
     patch('fix_optimization.patch', when='@2.0.2.0:%fj')
 
-    def configure_args(self):
+
+    depends_on('zlib')
+    depends_on('zstd')
+
+    extends('python', when="+python")
+    depends_on('python@3:', when="+python")
+    depends_on('swig@3.12:', when="+python")
+    depends_on('py-numpy', when="+python")
+    depends_on('hdf5', when="+hdf5")
+    depends_on('netcdf-c', when="+netcdf")
+
+    #before 2.1.8.1 this was a autotools package
+    @when("@:2.1.8.0")
+    def build(self, spec, prefix):
+        pass
```

Comment:
I would remove this, Autotools still has build and install phases we can take advantage of.

Reply:
When I asked @becker33 about the best way to leave the older versions of the package to keep using autotools and the newer versions to use cmake, he pointed me to the Nest package which does this.

It also fails to build if I don't do this.

Reply:
> ... the Nest package which does this.

Personally I think the `nest` package should also be changed.

> It also fails to build if I don't do this.

Can you elaborate on this?

Reply:
You get an error in python thrown from within spack that the files haven't been staged.

Reply:
Can you run with `spack -d install sz` and post the error message here? This should definitely work.

Reply:
```console
==> [2020-02-21-13:55:49.349572] Installing sz
==> [2020-02-21-13:55:49.349605] Searching for binary cache of sz
==> [2020-02-21-13:55:49.349775] No Spack mirrors are currently configured
==> [2020-02-21-13:55:49.350868] No binary for sz found: installing from source
==> [2020-02-21-13:55:49.350947] READ LOCK: /home/robertu/git/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> [2020-02-21-13:55:49.351719] READ LOCK: /home/robertu/git/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
==> [2020-02-21-13:55:49.366167] READ LOCK: /home/robertu/git/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> [2020-02-21-13:55:49.421275] [TOOLCHAIN] MATCH Gcc
==> [2020-02-21-13:55:49.430624] '/software/gcc/8.2.0/bin/gcc' '-v' '/local_scratch/pbs.8144260.pbs02/spack-implicit-link-infodVXUOI/main.c' '-o' '/local_scratch/pbs.8144260.pbs02/spack-implicit-link-infodVXUOI/output'
==> [2020-02-21-13:55:49.482362] parsing implicit link info
==> [2020-02-21-13:55:49.482652] linker line:  /software/gcc/8.2.0/libexec/gcc/x86_64-pc-linux-gnu/8.2.0/collect2 -plugin /software/gcc/8.2.0/libexec/gcc/x86_64-pc-linux-gnu/8.2.0/liblto_plugin.so -plugin-opt=/software/gcc/8.2.0/libexec/gcc/x86_64-pc-linux-gnu/8.2.0/lto-wrapper -plugin-opt=-fresolution=/local_scratch/pbs.8144260.pbs02/ccJVmSQ4.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o /local_scratch/pbs.8144260.pbs02/spack-implicit-link-infodVXUOI/output /lib/../lib64/crt1.o /lib/../lib64/crti.o /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/crtbegin.o -L/software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0 -L/software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/../../.. /local_scratch/pbs.8144260.pbs02/ccXAlOvh.o -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/crtend.o /lib/../lib64/crtn.o
==> [2020-02-21-13:55:49.482751] linkdir: /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0
==> [2020-02-21-13:55:49.482787] linkdir: /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/../../../../lib64
==> [2020-02-21-13:55:49.482818] linkdir: /lib/../lib64
==> [2020-02-21-13:55:49.482844] linkdir: /usr/lib/../lib64
==> [2020-02-21-13:55:49.482867] linkdir: /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/../../..
==> [2020-02-21-13:55:49.482913] found raw link dirs: /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0, /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/../../../../lib64, /lib/../lib64, /usr/lib/../lib64, /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/../../..
==> [2020-02-21-13:55:49.482980] found link dirs: /software/gcc/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0, /software/gcc/8.2.0/lib64, /lib64, /usr/lib64, /software/gcc/8.2.0/lib
==> [2020-02-21-13:55:49.535064] A dependency has updated CPATH, this may lead pkg-config to assume that the package is part of the system includes and omit it when invoked with '--cflags'.
==> [2020-02-21-13:55:49.553654] Using cached archive: /home/robertu/git/spack/var/spack/cache/_source-cache/archive/17/176c65b421bdec8e91010ffbc9c7bf7852c799972101d6b66d2a30d9702e59b0.tar.gz
==> [2020-02-21-13:55:49.563541] Staging archive: /local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/v2.0.2.0.tar.gz
==> [2020-02-21-13:55:49.563988] '/bin/tar' '-xf' '/local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/v2.0.2.0.tar.gz'
==> [2020-02-21-13:55:49.660461] Created stage in /local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay
==> [2020-02-21-13:55:49.660802] No patches needed for sz
==> [2020-02-21-13:55:49.660844] Building sz [CMakePackage]
==> [2020-02-21-13:55:49.660918] WRITE LOCK: /local_scratch/pbs.8144260.pbs02/robertu/spack-stage/.lock[4668049073539566329:1] [Acquiring]
==> [2020-02-21-13:55:49.661057] WRITE LOCK: /local_scratch/pbs.8144260.pbs02/robertu/spack-stage/.lock[4668049073539566329:1] [Acquired]
==> [2020-02-21-13:55:49.661169] WRITE LOCK: /home/robertu/git/spack/opt/spack/.spack-db/prefix_lock[4307824368065044079:1] [Acquiring]
==> [2020-02-21-13:55:49.662283] WRITE LOCK: /home/robertu/git/spack/opt/spack/.spack-db/prefix_lock[4307824368065044079:1] [Acquired]
==> [2020-02-21-13:55:49.667773] Executing phase: 'cmake'
==> [2020-02-21-13:55:53.815384] Executing phase: 'build'
==> [2020-02-21-13:55:53.817534] WRITE LOCK: /home/robertu/git/spack/opt/spack/.spack-db/prefix_lock[4307824368065044079:1] [Released]
==> [2020-02-21-13:55:53.817870] WRITE LOCK: /local_scratch/pbs.8144260.pbs02/robertu/spack-stage/.lock[4668049073539566329:1] [Released]
==> [2020-02-21-13:55:53.821141] Error: OSError: [Errno 2] No such file or directory: '/local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/spack-build'

/home/robertu/git/spack/lib/spack/spack/build_systems/cmake.py:245, in build:
        244    def build(self, spec, prefix):
  >>    245        """Make the build targets"""
        246        with working_dir(self.build_directory):
        247            if self.generator == 'Unix Makefiles':
        248                inspect.getmodule(self).make(*self.build_targets)

See build log for details:
  /local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/spack-build-out.txt

Traceback (most recent call last):
  File "/home/robertu/git/spack/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/home/robertu/git/spack/lib/spack/spack/package.py", line 1751, in build_process
    phase(self.spec, self.prefix)
  File "/home/robertu/git/spack/lib/spack/spack/package.py", line 111, in phase_wrapper
    phase(spec, prefix)
  File "/home/robertu/git/spack/lib/spack/spack/build_systems/cmake.py", line 246, in build
    with working_dir(self.build_directory):
  File "/usr/lib64/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/robertu/git/spack/lib/spack/llnl/util/filesystem.py", line 603, in working_dir
    os.chdir(dirname)
OSError: [Errno 2] No such file or directory: '/local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/spack-build'
==> [2020-02-21-13:55:53.821521] READ LOCK: /home/robertu/git/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> [2020-02-21-13:55:53.822583] READ LOCK: /home/robertu/git/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
==> [2020-02-21-13:55:53.840591] READ LOCK: /home/robertu/git/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> [2020-02-21-13:55:53.843684] ChildError: OSError: [Errno 2] No such file or directory: '/local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/spack-build'
/home/robertu/git/spack/lib/spack/spack/build_systems/cmake.py:245, in build:
        244    def build(self, spec, prefix):
  >>    245        """Make the build targets"""
        246        with working_dir(self.build_directory):
        247            if self.generator == 'Unix Makefiles':
        248                inspect.getmodule(self).make(*self.build_targets)

See build log for details:
  /local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/spack-build-out.txt
Traceback (most recent call last):
  File "/home/robertu/git/spack/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/home/robertu/git/spack/lib/spack/spack/package.py", line 1751, in build_process
    phase(self.spec, self.prefix)
  File "/home/robertu/git/spack/lib/spack/spack/package.py", line 111, in phase_wrapper
    phase(spec, prefix)
  File "/home/robertu/git/spack/lib/spack/spack/build_systems/cmake.py", line 246, in build
    with working_dir(self.build_directory):
  File "/usr/lib64/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/robertu/git/spack/lib/spack/llnl/util/filesystem.py", line 603, in working_dir
    os.chdir(dirname)
OSError: [Errno 2] No such file or directory: '/local_scratch/pbs.8144260.pbs02/robertu/spack-stage/spack-stage-sz-2.0.2.0-o6io5yf2so6n5a34kknpqnlwm2dwpgay/spack-build'
```

Reply:
This only affects the versions built with autotools.

Reply:
Ah, the issue is that `CMakePackage.build_directory` defaults to `spack-build`, while `AutotoolsPackage.build_directory` defaults to `.`. So if you add:
```python
build_directory = 'spack-build'
```
it will build in a subdirectory for both Autotools and CMake. You can set this based on version if Autotools doesn't like building in a subdirectory.

Reply:
How do I set this based on version?  If I try to set it in `cmake()` for the the old version to `'.'` I get an AttributeError

Reply:
Don't set it in `cmake()`, set it at the class-level.

Reply:
and use the `@when` annotation above it?

Reply:
If you do need to set it by version, it would look like:
```python
@property
def build_directory(self):
    if self.version >= Version('2.1.8.1'):
        return 'spack-build'
    else:
        return '.'
```

Reply:
But first try:
```python
build_directory = 'spack-build'
```
and see if we can always build in a subdirectory for both Autotools and CMake.

Reply:
Sheng did some crazy things in his autotools files, and without the property it doesn't work.